core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -264,12 +264,37 @@ class TaskMetrics private[spark] () extends Serializable {
   /**
    * External accumulators registered with this task.
    */
-  @transient private[spark] lazy val _externalAccums = new CopyOnWriteArrayList[AccumulatorV2[_, _]]
+  @transient private[spark] lazy val _externalAccums = new ArrayBuffer[AccumulatorV2[_, _]]
 
-  private[spark] def externalAccums = _externalAccums.asScala
+  private[spark] def externalAccums: Seq[AccumulatorV2[_, _]] = withReadLock {
+    _externalAccums.toSeq
```

Comment:
can we double check if `ArrayBuffer#toSeq` is cheap?

Reply:
It's cheap. It essentially creates a linked list without clone items. The complexity is O(n).

Reply:
In Scala 2.13, `ArrayBuffer.toSeq` ultimately invokes `Nil.prependedAll`, and constructs a new `immutable.List` by iterating over the `ArrayBuffer`.

Reply:
Since it compiles, I think there is no caller of `externalAccums` trying to modify it. I think it's OK to eliminate the `toSeq`.

Reply:
It might still be necessary to convert to the new `immutable.Seq`. The initial change of `_externalAccums` to use `CopyOnWriteArrayList` was to address the bug described in SPARK-39696, at which time the following test case was added:

https://github.com/apache/spark/blob/d5dc22356d23b9d15b4f7c195e60c3767ea205ad/core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala#L274-L293

IIRC, from a test a long time ago, converting `ArrayBuffer` to an `Array` first and then wrapping it as an `ArraySeq` showed better performance than using `toSeq` directly

```scala
import org.apache.spark.util.ArrayImplicits._
_externalAccums.toArray.toImmutableArraySeq
```

Reply:
Agree, we should ensure this is thread safe.

Reply:
Updated to `_externalAccums.toArray.toImmutableArraySeq`.